README.md
![hackneyed-light-preview](preview-Hackneyed.png "The sands of time are running out for you, bro")
![hackneyed-dark-preview](preview-Hackneyed-Dark.png "SO PUUUUURE! SO COOOOOLD!")

Overview
--------

Hackneyed is a scalable cursor theme mildly resembing old Windows 3.x/NT 3.x cursors. Dark and left-handed versions are also available.


Building
--------
Minimum dependencies:

* ImageMagick (>=6.8.6 but not 7.0, as it has huge API changes)
* Inkscape (>=0.92.3)
* GNU `make` and `bash`
* `xcursorgen` (part of `xorg-x11-apps`)

* To generate and pack Windows cursors:
	* A working C compiler;
	* `ImageMagick` >= 6.8.6;
	* `libpng` >= 1.6.36;
	* `pkg-config`;
	* `zip`.

Production targets (all of them generate tarballs for distribution):

* `dist`: build the scalable, right-handed theme;
* `dist.left`: build the scalable, left-handed theme;
* `dist.<size>`, where `size` can be small, medium or large: build a fixed-size, right-handed theme of the specified size;
* `dist.<size>.left`: build a fixed-size, left-handed theme of the specified size;
* `windows-cursors` for Windows. These cursors employ an alpha channel and are compatible with Windows XP and later (Windows 10 users: they are high-DPI ready as well).

`make all-dist` targets all of the above (except `windows-cursors`), including all fixed-size themes in all available sizes.

`small`, `medium` and `large` are currently defined as 24, 36 and 48, respectively. The making of individual cursors recognizes the integer argument only (see below).

The targets described below are meant for debugging and do not generate tarballs:

* `theme`: make the scalable, right-handed theme;
* `theme.left`: make the scalable, right-handed theme;
* `theme.<size>`: make a fixed-size, right-handed theme of the specified size;
* `theme.<size>.left`: make a fixed-size, left-handed theme of the specified size.

All of them run `do-symlinks.sh` when finished. The targets below do not:

* `all`: make the scalable, right-handed theme;
* `all.<size>`: make a fixed-size, right-handed theme of the specified size;
* `all.left`: make the scalable, left-handed theme;
* `all.<size>.left`: make a fixed-size, left-handed theme of the specified size.

Individual cursors can be made with `make <cursor_name>.<size in pixels>.<orientation>`, e.g.:

* `make default.24.left` for a left-handed, 24px `default` cursor;
* `make default.24` for a right-handed, 24px cursor;
* `make default.left` for a scalable, left-handed cursor;
* `make default.24.png` for a 24px PNG only. PNG sizes aren't hardcoded, so you can specify any size you want;
* or simply `make default` for a scalable, right-handed cursor.

Parallel jobs (`-j`) are recommended (Inkscape is _slow_).

If you don't feel like building it from source, grab the latet builds from the artwork page on [openDesktop.org](https://www.opendesktop.org/p/999998/) or from [Gitlab](https://gitlab.com/Enthymeme/hackneyed-x11-cursors/-/releases).

To build the dark theme, tell `make` to use the dark versions of `COMMON_SOURCE`, `RSVG_SOURCE` and `LSVG_SOURCE`, and change `THEME_NAME`:

`$ make -B THEME_NAME=Hackneyed-Dark COMMON_SOURCE=theme/common-dark.svg RSVG_SOURCE=theme/right-handed-dark.svg LSVG_SOURCE=theme/left-handed-dark.svg <target>`


Installation
------------
Simply choose a tarball and extract it to `$HOME/.icons`. If you've built it from source, pick a size or variant (e.g., `Hackneyed`, `LHackneyed-36px`) and move it to `$HOME/.icons`.


License
-------
Hackneyed is released under the MIT/X11 license.


Credits
-------
* do-symlinks.sh, rewrite of addmissing.sh, taken from [dummy-x11-cursors](https://www.opendesktop.org/p/999853/) by ultr

* some SVGs taken from openclipart.org (pencil, pirate and coffee_mug as far as I can remember)

* pencil and pirate borrowed from Breeze

* monolithic SVG idea (and the SVG itself) taken from KDE's [Breeze theme](https://github.com/KDE/breeze/tree/master/cursors) by Ken Vermette, who probably liked Jakub Steiner's DMZ way too much (but not enough to keep the Python script)

* ico2cur.c, a C rendition of [ico2cur.py](https://gist.github.com/RyanBalfanz/2371463)


Bugs/enhancements
-----------------
Suggestions and bug reports can be made here or on openDesktop.


A word about hashes
-------------------
libXcursor-based applications will return a hash for custom cursors when you run them
like `env XCURSOR_DISCOVER=1 <executable>` on a terminal. Take into account only the hashes
that return zero, e.g., "Cursor hash `0xdeadbeef` returns `0x0`".

(Forget about doing this to animated cursors: a hash is generated for every frame.)


Useful links
------------
* [Freedesktop.org's cursor specification](http://www.freedesktop.org/wiki/Specifications/cursor-spec/)

* [CSS cursors from W3C](http://dev.w3.org/csswg/css-ui/#propdef-cursor "2drafty4u")

* [Mozilla's test page for CSS cursors](https://developer.mozilla.org/en-US/docs/Web/CSS/cursor)

* [Qt requirements for X11 cursors](http://doc.qt.io/qt-5/qcursor.html#a-note-for-x11-users)

* [Core X11 cursors](http://tronche.com/gui/x/xlib/appendix/b/ "coffee_mug > all")


Setting a default theme for all applications
--------------------------------------------
Setting the cursor theme through a graphical interface doesn't always propagate to all applications. _Most of the time_, this can be solved by creating a file called `index.theme` with the following content...

```
[Icon Theme]
inherits=Hackneyed      # or Hackneyed-48px, LHackneyed etc.
```

...and saving it in `~/.icons/default` (create the folder if it doesn't exist already).

Chromium and Chromium-based browsers won't pick up a new theme unless you restart the browser (seems to be fixed in Chromium 80).


Known bugs
----------
The bugs described below are NotMyProblem<sup>TM</sup> and should be dealt with by upstream:

- **Qt on Wayland outside Plasma:** If you're not on KDE, create a `~/.icons/default/index.theme` as described above or use `XCURSOR_THEME=Hackneyed XCURSOR_SIZE=<current cursor size> QT_QPA_PLATFORM=wayland <qt application>` and check if something is amiss--as of now, Qt's (hopefully) makeshift CSD implementation switches to `split_h` and `split_v` in the borders of the window, which is *wrong*. At least within KDE itself, cursors use the proper size and shape.

- Qt on Wayland outside KDE is also not *that* great if you want to resize a window without holding `<Super> + middle click` *every goddamn time*, as having no "virtual" border to grab onto will be a test of patience and pointing device skills.

- **`all-scroll` as `move` on Firefox, Chromium 62+ and derivatives:** Firefox has a [14-year-old bug](https://bugzilla.mozilla.org/show_bug.cgi?id=275174) concerning exactly this. Still nothing about Chromium.

- **GTK 4.** GTK 4 only looks for cursor themes in `/usr/local/share/themes` or `/usr/share/themes`. If Flatpak is installed, then it searches whatever `<datadir>/flatpak/exports/share` is available (either system-wide ones or in `~/.local/share/flatpak`). I really hope this was a mistake on their part.

To do
-----
Hackneyed's "build system" sucks and I should replace it eventually.


CHANGELOG
0.8.2
- fix 'arrow' symlink (must point to 'default')

0.8.1
- support Inkscape 1.0 (huge reduction in build times)
- merge make-ani-frames.sh and make-png.sh
- rewrite large sections of the Makefile to make it POSIX compatible
- rewrite png2cur to output bitmap cursors by default. in some situations,
  Windows misidentifies the 32px PNG as a monochrome bitmap (can't be arsed
  to track which build of Windows 10 started doing this)

0.8:
- *_arrow -> sb_*_arrow
- rewrite do-symlinks.sh
- move symlink definitions to a separate file
- review all symlinks as a result
- fix brokenness on animaker.c (uninitialized variable) and png2cur.c (flawed logic
  while testing for sizes with no hotspot) which means I was shipping broken cursors
  for Windows. thankfully nobody uses them besides me

0.7.4:
- redraw pirate, exchange and color-picker

0.7.3:
- all-scroll: don't reverse colors on the center dot

0.7.2:
- revert pointer to a standard pointing hand
- get pirate and pencil from Breeze
- minor misaligment fixes
- double down on the horn hand sign for wayland-cursor

0.7.1:
- enlarge hands by a few pixels, they were way too tiny

0.7:
- actually put effort into aligning to pixels
- redo mostly everything as a result
- steal pointer from X11, steal zoom again from Breeze
- dark theme ofc

0.6.1
- progress: animate hourglass
- wait: redo animation because it sucked
- reinstate make-png.sh
- make-animated-cursor.sh can now deal with variants (i.e. left-handed
  pointers)

0.6
- new size set (24-36-48-60-72), more consistency across sizes
- overall optimizations to account for the smaller size
- redo static/animated hourglass
- drop wc-large.svg, use a 32px png made on the fly
- rewrite Makefile to be more maintainable

0.5.3.1
- fix all-dist targets copying extra files

0.5.3
- no comments

0.5.2
- make single-sized themes (for desktops like LXDE, and XFCE when size > 48)

0.5.1
- symlink grab -> openhand

0.5
- overhaul the artwork entirely
- SVGs now amount to two, one for each variation; individual cursors
  are exported through ids with inkscape
- ditch all but one of the IM hacks

0.4
- text, crosshair: avoid blurry shapes by aligning to pixel boundaries
- pencil, draft: recoloring and reshaping
- pointer, openhand, closedhand: be a bit more anatomically correct, align
  to pixels as well so that the fingers don't look weird
- zoom: make it fully white to stand out more
- zoom-in, zoom-out: change colors
- context-menu: the menu now looks more like a menu
- overall repositioning of overlays because I must have fucked up with
  something

0.3.26
- text, crosshair: more elegant shapes

0.3.24
- make zoom's lens (and consequently zoom-in and zoom-out's) slightly bigger

0.3.23
- swap back col-resize/row-resize, GTK+ applications have fixed themselves

0.3.22
- left-handed versions of openhand, closedhand and pointer

0.3.21.1
- relicense under the MIT/X11 license

0.3.21
- pointer: actually fix the length of the index finger

0.3.20
- soften shadows once again;
- fix the length of the finger in pointer;
- add draped_box;
- widen the stem of text by 1 pixel to avoid blurriness

0.3.19
- added some legacy cursors (based_arrow_*, *_tee, *_angle)

0.3.18
- give some life to draft by coloring it, make pencil look like it;
- make shadows more consistent across pointers, specially the hands;
- change colors of dnd-move's fleur and the question mark in help,
  so they are more noticeable;
- rework the hourglass to make it look better in small sizes;
- wayland-cursor is now a sign of the horns (that was on the previous
  version, but I forgot to write it and I don't quite like historical
  revisionism)

0.3.17
- turn right_ptr into a horizontal arrow pointing right, so
  that it looks distinct in the left-handed theme;
- optimize context-menu to look good in small sizes;
- add draft, wayland-cursor and other arrows, as found in
  the new KDE cursor theme;
- start writing changelogs

0.3.16:
- added dnd-move and dnd-no-drop;
- softened shadows for tcross and xterm/ibeam;
- speaking of tcross, it's less ugly now;
- lots of debugging to figure out where those symlinks should
  really point to. Updates will become less frequent from now on,
  unless a bug is found or someone makes a request

0.3.15:
- fix col- and row-resize mess (I blame my sources)

0.3.14:
- Add more sizes because all the cool kids are doing it, and
  I want to be cool.

0.3.13:
- cursors for the young and old

0.3.12:
- left-handed version

0.3.9:
- add cursors that were still missing
- rework it some more because it's never good enough;
- replace the ugly fountain pen with a more distinct pencil

0.3.6:
- small changes to make it more consistent

0.3.4:
- new cursors added

0.3.2:
- hotspot fixes
- add animation for 48x48 hourglass as well

0.2
- fix size of the default pointer


build-quicker.sh
#!/bin/sh
source $stdenv/setup
COMMON="text vertical-text sb_up_arrow sb_down_arrow sb_left_arrow sb_right_arrow pirate based_arrow_up based_arrow_down ns-resize ew-resize nwse-resize nesw-resize move plus crosshair n-resize s-resize w-resize e-resize nw-resize ne-resize sw-resize se-resize center_ptr all-scroll not-allowed draped_box wayland-cursor X_cursor ul_angle ur_angle ll_angle lr_angle split_v split_h top_tee bottom_tee left_tee right_tee"
CURSORS="alias color-picker context-menu copy default help pencil dnd-move zoom zoom-in zoom-out no-drop draft right_ptr openhand closedhand pointer coffee_mug exchange"
smallest=24
builddir=build-test
outdir=out
mkdir -p $builddir/in
mkdir -p $outdir
for x in $CURSORS $COMMON
do cat theme/*/$x.in > $builddir/in/$x.in
done
awk -vbuilddir=$builddir '{ name = $4; sub(/\.[0-9]+\.png/, "", name); for (i = 1; i <= 16; ++i) printf "%u %u %u %s-%u.%u.png %u\n", $1, $2, $3, name, i, $1, i == 5 ? 300 : 30 > builddir"/in/"name".in" }' theme/*/wait.in theme/*/progress.in
for theme in white dark
do mkdir -p $builddir/$theme
mkdir -p $outdir/$theme/cursors
if [ "$theme" = white ]
then fancy_name="Hackneyed-White"
else fancy_name="Hackneyed-Dark"
fi
for s in 24 36 48 60 72
do for x in $CURSORS
do printf 'theme/right-handed-%s.svg -d %u -i %s -e %s/%s.%u.png\n' $theme $((96 * s / smallest)) $x $builddir/$theme $x $s
done
for x in $COMMON
do printf 'theme/common-%s.svg -d %u -i %s -e %s/%s.%u.png\n' $theme $((96 * s / smallest)) $x $builddir/$theme $x $s
done
for i in $(seq 1 16)
do printf 'theme/common-%s.svg -d %u -i wait-%u -e %s/wait-%u.%u.png\n' $theme $((96 * s / smallest)) $i $builddir/$theme $i $s
printf 'theme/right-handed-%s.svg -d %u -i progress-%u -e %s/progress-%u.%u.png\n' $theme $((96 * s / smallest)) $i $builddir/$theme $i $s
done
done | $INKSCAPE --shell
for x in $CURSORS $COMMON wait progress
do $XCURSORGEN -p $builddir/$theme $builddir/in/$x.in $outdir/$theme/cursors/$x
done
while IFS="$(printf '\t')" read x ys
do for y in $ys
do ln -Lsr $outdir/$theme/cursors/$x $outdir/$theme/cursors/$y || true
done
done < theme/symlinks
cat <<EOF > $outdir/$theme/index.theme
[Icon Theme]
Name = $fancy_name (scalable)
Comment = Windows 3.x-inspired cursors
Example = progress
EOF
mkdir -p $out/share/icons/$fancy_name
cp -r $outdir/$theme/* $out/share/icons/$fancy_name
done
